Fix: email signup logic

BerkeleyTrue opened the pull request:
closes #16288 
- [x] Create new AuthToken
- [x] GitHub auth should not overwrite email.

QA:
- Create new account
  - get new account message
  - get new account email
  - follow new account email link to signin
- Sign in with old account
  - get sign in message
  - get sign in email
  - follow sign in email to signin

BerkeleyTrue committed:
fix(models/User): Remove dup before create

BerkeleyTrue committed:
fix(server/user): End signin middleware early if signed in

BerkeleyTrue committed:
chore(packages): Update validator and express-validator

BerkeleyTrue committed:
fix(models): Move custom method to loopback model extension

BerkeleyTrue committed:
refactor(server/authenticate): Move auth api into own boot script

separate from user account/display methods

BerkeleyTrue committed:
refactor(User): Make `requestAuthEmail` into instance method

raisedadead commented on common/models/user.json:
```diff
@@ -22,6 +22,10 @@
       "type": "date"
     },
+    "emailVerified": {
+      "type": "bolean",
```
```diff
-      "type": "bolean",
+      "type": "boolean",
```

BerkeleyTrue replied:
fixed

raisedadead commented:
Hi @BerkeleyTrue this is an excellent direction IMHO, especially moving the logic to the dedicated boot script.

I know this is super busy time for you, but could you share any of the error logs that you saw in production (or analysis of it?).

I suspect and fear that the accessToken Objects and the User Objects were not correctly linked? Just curious and concerned to find the root cause.

BerkeleyTrue committed:
fix(passwordless): Reduce db calls, run in parallel

Adds validations, reduces the number of database calls, separates
concers. reduces logic

BerkeleyTrue committed:
fix(User): Typo

BerkeleyTrue commented:
@raisedadead I suspect the issue lies with some code running currently. There where also a great many db calls being made. The errors I where seeing where `cannot set headers after they are set`, which is express' way of telling us that the request has been fulfilled but some async process is attempting to the send more data. 

I didn't grab screen caps of these errors but I'd seen them enough to know that there was at the minimum an async issue.

BerkeleyTrue committed:
fix(User.updateEmail): Reduce code logic. defer promises

raisedadead commented:
> I suspect the issue lies with some code running currently. There where also a great many db calls being made.

Yup, the initial implementation relied on nested calls to db. I admit that it was way too procedural.

I don't remember now, but I guess the flexibility was for accounts which may not have had the `emailVerified` property set (for example, old accounts, GitHub created accounts, etc).

But, now we are enforcing only one path to create/login via email (which is good). Hence this approach suits better.

I'll definitely want to keep in mind though of the behaviour, if User Objects missing this `emailVerified` property.

Will test when this is ready, thanks for the refactor.

BerkeleyTrue committed:
feat(User): Refactor wait period logic

BerkeleyTrue committed:
fix(User.requestAuthEmail): Conditionally check for email availibility

Only check if email is available if it is not already associated with
this account

BerkeleyTrue committed:
fix(AccessToken): Move extensions to boot script

Loopback does not facilitate built in model extensions in the same way
as user defined models

BerkeleyTrue committed:
fix(User): Email is off the constructor

BerkeleyTrue committed:
fix(boot/auth): Fix typo

BerkeleyTrue committed:
fix(middlewares/errorhandler): Remove npm module and refactor for fcc use case

BerkeleyTrue committed:
fix(passport/login): Use middleware instead of bootscript

prototype overwrite doesn't seem have any effect

BerkeleyTrue committed:
fix(User): Fix typos

BerkeleyTrue committed:
chore(gulpfile): Reduce wepback log level to info

BerkeleyTrue committed:
fix(handled-errors): Handled errors should default to 200

BerkeleyTrue committed:
fix(User): Unify old remove save hook with before save

Since we are no longer using remote api to create accounts those hooks
are useless. Unifying them will ensure they apply to all
save/update/creates

BerkeleyTrue committed:
fix(User): Add isSignUp logic

emailVerfied field no longer indicates a new user.

BerkeleyTrue committed:
fix(User): Deny old auth api

BerkeleyTrue commented:
This is almost ready. I'd like to make one more change, adding a new model based of the accessToken called authToken. This ensures that there is a complete separation between the two types of tokens

raisedadead commented:
Nice work @BerkeleyTrue !

>  I'd like to make one more change, adding a new model based of the accessToken called authToken. This ensures that there is a complete separation between the two types of tokens

Sounds reasonable to me, but with that we need to define the lifecycle methods on that too.

What I mean is we would have to implement the:
```
this.createAccessToken$({ ttl: 15 * 60 * 1000 });
```

and its associated lifecycle methods?
If not a priority we should defer it to later?

BerkeleyTrue committed:
chore(Loopback): Normalize configs using cli

BerkeleyTrue committed:
feat(User/Auth): Use stand alone auth token

BerkeleyTrue committed:
fix(User): Allow confirm

This is still used for email confirmation on update

BerkeleyTrue committed:
fix(User.confirm): Overwrite confirm w/ custom method

Reduce db calls, implement old logic

BerkeleyTrue committed:
fix(User): Move login logic into user model

raisedadead commented:
WIP - QA

Testing locally
QA:
- [x] Create new account
  - [x] get new account message
  - [x] get new account email
  - [x] follow new account email link to signin
- [ ] Sign in with old account
  - [ ] get sign in message
  - [ ] get sign in email
  - [ ] follow sign in email to signin

Observations:

- The TTL for the AccessToken Object (long term) is `NaN`
```
{
    "_id" : "PwtRSaZPbcvver5t9E0gks7xASZLq4o2znvYAqB903lSiRbbdR4wo8oUU6DUyIhu",
    "ttl" : NaN,
    "created" : ISODate("2017-12-30T07:36:48.123Z"),
    "userId" : ObjectId("5a47420bf3f3e8298faaa18f")
}
```

raisedadead commented:
Additional Notes (to be taken in new PR, since they are not bugs per say because of changes):

-  The update email path breaks with a `Error: Authorization Required`
  -  Email is sent and received correctly.
  -  Clicking on the email is hitting LoopBack's native implementation (`/confirm` endpoint) and needs to be rather implemented now.

raisedadead commented:
@BerkeleyTrue 

Everything looks good to me.

One thing to handle though is the destruction of the old token as I am able to do repeated logins after sign out for second time on using the same previous Login Requested email.

Steps:
1. Create new account / Logout
2. Sign in with old account (from step 1) / Logout
3. Click the same (from step 2) / Logout
4. Repeat steps 2-3.

BerkeleyTrue commented:
@raisedadead Sorry, I left three commits on my local. It should fix those problems.

BerkeleyTrue committed:
feat(boot/auth): Add signup disabled debug info

BerkeleyTrue committed:
fix(Nav): My Profile => Settings

BerkeleyTrue committed:
fix(Auth): Error type in json payload

BerkeleyTrue committed:
fix(views/signup): Error should be danger

raisedadead commented:
Thanks @BerkeleyTrue, Happy new year. I'll QA tonight and merge.

raisedadead commented:
#### :sparkles: :sparkles: **WOOT!** :sparkles: :sparkles: 

Thanks @BerkeleyTrue, Nice work! Everything works like a charm. 
Lets get this on beta, and test the response times in the wild.
